### PR TITLE
Add sanity test for multi DC and several improvements for test.py

### PR DESCRIPTION
**Pull request by xtrey**
Add sanity tests for multi dc

Fix writing cassandra-rackdc.properties with correct format data instead of yaml
Add a parameter to overwrite RF for specific DC
Add the possibility to connect cql to the specific node

In this PR 4 tests were added to test multi-DC functionality. One is added from initial commit were multi-DC possibility were introduced, however, this test was not commited. Three of them are migrations from dtest, that later will be deleted. To be able to execute migrated tests additional functionality is added: the ability to connect cql to the specific node in the cluster instead of pooled connection and the possibility to overwrite the replication factor for the specific DC. To be able to use the multi DC in test.py issue with the incorrect format of the properties file fixed in this PR.

### CODE REVIEW

File: test/topology_custom/test_multidc.py
Comment (enaydanov) [inline test/topology_custom/test_multidc.py:13]: These are libraries from the stdlib, they should be in a first block separated from the 3rd-party libraries by one empty line.

File: test/topology_custom/test_multidc.py
Comment (enaydanov) [inline test/topology_custom/test_multidc.py:83]: wdyt about following loop:

```python
for i in [1, 1, 2, 2]:
    ...
```

File: test/topology_custom/test_multidc.py
Comment (enaydanov) [inline test/topology_custom/test_multidc.py:143]: Need to add some explanation about records you put to DB here.

File: test/topology_custom/test_multidc.py
Comment (enaydanov) [inline test/topology_custom/test_multidc.py:179]: Without any comments it looks like some magic : )

**Review by enaydanov — CHANGES_REQUESTED** [review]
In addition:
1. Why you need 2 commits in this PR?
2. Please, add description to the commit/commits and to the PR
3. Add more logging in tests.

**Comment by xtrey** [comment]
> In addition:
> 
> 1. Why you need 2 commits in this PR?
> 2. Please, add description to the commit/commits and to the PR
> 3. Add more logging in tests.

1. No, there is no need in several commit. Fixed
2. Added more information
3. Additional logging added.

**Comment by yaronkaikov** [comment]
@xtrey @enaydanov i am running the CI again https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7143/ this time it will run on fleet (we have different value for `aio-max-nr` ) , let see if it pass, i will align all local nodes as well

**Comment by xtrey** [comment]
@yaronkaikov thanks. Looks like change is helped. Only one test failed, I'll investigate it.

File: test/pylib/random_tables.py
Comment (nyh) [inline test/pylib/random_tables.py:266]: Why do you want to put this variable in "self"?  Do you plan to use it anywhere except in this function?

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:7]: I don't understand the relevance of this comment. Maybe it's a copypasto?

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:30]: It looks like you are adding now "a new node", but 30 of them. 
Is there any benefit to adding 30 and not 10 or 3?

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:26]: This comment doesn't really explain what this test is about. First of all, it's not just a "mult-dc environment", it's a particular setup where, if I understand correctly, you have 30 nodes in 30 different DCs.
Second, what does this test actually check? That the 30 nodes are able to boot (how is boot different in the case of 30 different CDs and one DC)? That you are able to create a table in a 30-DC setup? I don't see any data operations below, only the table creation, so maybe you can explain why it's interesting.

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:47]: https://github.com/scylladb/scylladb/issues/3539 doesn't seem related. So which issue did you mean?

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:73]: Nitpick: It seems this test is identical to the previous one except for the array, so you could have done one copy of the test with "pytest.parameterize".

Another benefit is that you won't need the extra putget() function (which I, as a reader, have absolutely no idea what it does), you can insert it inside this function and let the reader read the test sequentially without jumping around the file.

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:102]: Another thing you forgot to mention in the commit message is what tests you are adding, and why. This particular test seems a direct "borrowing" of the existing dtest multidc_additional_tests.py. I think it was worth mentioning. And it may be worth considering (please do!) deleting this dtest later - we don't need exactly the same test twice.

Are other tests added here also translations of existing dtests?

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:128]: I know this test is copied from the existin dtest, so I can't complain, but I'm still curious why it was important to query DC1 with CL=ONE... If the bug we were hunting was about LOCAL_QUORUM on DC2, doesn't it make sense to also query DC1 with LOCAL_QUORUM?

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:139]: It seems the code below actually adds nb_keys * 100 rows?
What is nb_keys about at all?

Personally, I would love for this function to disappear. If I understand correctly, you only use it once in the code. So just insert it where its needed, and add a comment *exactly* what it is writing.

Also, is all this elaborate code to write 100 rows and then rewrite some of them, etc., needed for this test? Does it test anything that just writing 10 rows once wouldn't have tested?

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:177]: Again, I think (but it's a matter of style, I guess) that reading this code would have been much simpler without all these functions and just one linear code which writes a bunch of data, reads it back and checks if it has the right value. It will be a long function - but clearer, could be read in a linear fashion (one line after the other) and no need to name functions, parameters (like that nb_keys which is always 1 anyway), etc.

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:173]: So after the above generic function with an "nb_keys" parameter, you end up calling it just once, with nb_keys=1? :-( Can you get rid of this parameter and make everything easier to understand?

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:182]: Adding this assert_length_equal convenience function  is completely unnecessary and makes it harder to read the code. The idiomatic pytest way to do this is simply:
```
    assert len(res) == 100
```
And that's it! No need for an error string (assertion failures are very clear in pytest) and definitely not for a wrapper function for this assert.

**Review by nyh — CHANGES_REQUESTED** [review]

File: test/pylib/manager_client.py
Comment (xtrey) [reply test/pylib/manager_client.py:88]: Changed to CQL instead of CQLSH. 

Sure, will modify the commit message when I'll resolve all comments.

File: test/pylib/random_tables.py
Comment (xtrey) [reply test/pylib/random_tables.py:266]: The idea here just not to repeat the same query twice. It used later in modifying the the replication factor for each DC in lines 267-270

File: test/topology_custom/test_multidc.py
Comment (xtrey) [reply test/topology_custom/test_multidc.py:7]: Correct, this is a copypaste from the https://github.com/kostja/scylla/tree/multidc2. There was one test where @kostja  made the initial commit to support multi-DC. I'll delete this comment to remove confusion.

File: test/topology_custom/test_multidc.py
Comment (xtrey) [reply test/topology_custom/test_multidc.py:30]: This test is copypaste. I do believe the idea of this test is to check that ScyllaDB will be able to handle a large amount of DC. Each node in this test is added to its own DC, so the test generally checks the ability to work with 30 DC.

File: test/topology_custom/test_multidc.py
Comment (xtrey) [reply test/topology_custom/test_multidc.py:102]: All tests presented here are borrowed: 3 from dtests and 1 from https://github.com/kostja/scylla/blob/multidc2/test/topology_custom/test_multidc.py
Sure, I'll delete this tests from dtest once this PR will be merged. You are completely right there is no need to have the same tests in different repos.

File: test/topology_custom/test_multidc.py
Comment (xtrey) [reply test/topology_custom/test_multidc.py:73]: Good point. I'll rewrite it to be parametrized.

File: test/topology_custom/test_multidc.py
Comment (xtrey) [reply test/topology_custom/test_multidc.py]: Done

File: test/topology_custom/test_multidc.py
Comment (xtrey) [reply test/topology_custom/test_multidc.py]: Modified and simplified code.
This is a migrated test, and I think the idea behind it is to check that there are no issues with updating the values in a multi-DC environment.

File: test/topology_custom/test_multidc.py
Comment (xtrey) [reply test/topology_custom/test_multidc.py]: Removed unnecessary parameter

File: test/pylib/random_tables.py
Comment (nyh) [reply test/pylib/random_tables.py:266]: Yes, but all of these are **inside** the function, so this could have been a local variable, not an attribute of "self". Unless I'm missing something...

File: test/topology_custom/test_multidc.py
Comment (nyh) [reply test/topology_custom/test_multidc.py]: copypaste from what?
If the intention is to check a large number of DCs, this should be specified in the test's comment, and perhaps even the test's name (e.g., call it many_dc instead of multi_dc). But then you need to ask yourself, is 30 really many? What kind of datastructure might we have in Scylla which can deal with 10 DCs fine, but 30 DCs is already "many"?

File: test/pylib/random_tables.py
Comment (nyh) [reply test/pylib/random_tables.py:266]: By the way, @kostja maybe you want to review these changes to test/pylib so it doesn't "rot" in a way you don't like?

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:24]: Please move cluster_config to right above the place it is used, it will be clearer.

File: test/topology_custom/test_multidc.py
Comment (nyh) [inline test/topology_custom/test_multidc.py:49]: This comment is no longer accurate - it's just one of the two cluster_config options.

**Review by nyh — CHANGES_REQUESTED** [review]
I like this version much better, thanks. Left just a few minor nitpicks.

File: test/topology_custom/test_multidc.py
Comment (kostja) [reply test/topology_custom/test_multidc.py]: yes, the idea is to explore a rare dimension - 30-dc setup . If there is any quadratic or cubic logic in scylla from the number of DCs the test should uncover that. calling random_tables() would cover for DML and DDL in such a cluster. I don't know the actual run time of this test now, if it's too big, we could perhaps reduce the number of DCs.

File: test/pylib/random_tables.py
Comment (kostja) [reply test/pylib/random_tables.py:266]: I partly wrote them so I should be disqualified here. But I am OK if @xtrey submits the code that might "rot" a bit - he will bear the responsibility for the outcome, since I hope going forward he will be handling all test.py issues.
Thanks for pinging though.

**Review by nyh — APPROVED** [review]
There is still one silly question I had about why a certain variable had to be a class variable (instead of a local variable) that I didn't get an answer to. The rest of this PR looks good.
I see the CI is still running so I can't merge it yet anyway.

File: test/pylib/random_tables.py
Comment (xtrey) [reply test/pylib/random_tables.py:266]: @nyh sorry, misunderstood you. You are completely right about moving it to the local variable.

**Comment by xtrey** [comment]
> There is still one silly question I had about why a certain variable had to be a class variable (instead of a local variable) that I didn't get an answer to. The rest of this PR looks good. I see the CI is still running so I can't merge it yet anyway.

My apologies, missed your reply regarding this. Fixed.

File: test/topology_custom/test_multidc.py
Comment (bhalevy) [reply test/topology_custom/test_multidc.py]: 5 DCs is plenty, and atypical

Comment (mykaul) [reply test/topology_custom/test_multidc.py]: One of our largest customers uses 5 DCs...

**Review by tgrabiec — APPROVED** [review]

File: test/topology_custom/test_multidc.py
Comment (xtrey) [reply test/topology_custom/test_multidc.py]: Changed to 5.

**Comment by nyh** [comment]
@xtrey I already merged this PR two hours before you pushed newer versions. Just now it was even promoted to master. Please send an incremental patch of whatever you changed in your last push (I don't know what it is because you didn't write an explanation what you changed in the last push).

**Comment by kbr-scylla** [comment]
The new test is flaky.

```
>       dc1_connection = manager.connect_to_node(servers[0])

test/topology_custom/test_multidc.py:126: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/pylib/manager_client.py:111: in connect_to_node
    return CassandraCluster([server.ip_addr], execution_profiles={EXEC_PROFILE_DEFAULT: profile}).connect()
cassandra/cluster.py:1761: in cassandra.cluster.Cluster.connect
    ???
cassandra/cluster.py:1787: in cassandra.cluster.Cluster.connect
    ???
cassandra/cluster.py:1774: in cassandra.cluster.Cluster.connect
    ???
cassandra/cluster.py:3603: in cassandra.cluster.ControlConnection.connect
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.cluster.NoHostAvailable: ('Unable to connect to any servers', {'127.112.126.21:9042': OperationTimedOut('errors=None, last_host=None')})
```

It's probably because you did not set increased timeouts like we do e.g. in topology/conftest.py for `cluster_con`:
```python3
def cluster_con(hosts: List[IPAddress | EndPoint], port: int, use_ssl: bool, auth_provider=None, load_balancing_policy=RoundRobinPolicy()):
    """Create a CQL Cluster connection object according to configuration.
       It does not .connect() yet."""
    assert len(hosts) > 0, "python driver connection needs at least one host to connect to"
    profile = ExecutionProfile(
        load_balancing_policy=load_balancing_policy,
        consistency_level=ConsistencyLevel.LOCAL_QUORUM,
        serial_consistency_level=ConsistencyLevel.LOCAL_SERIAL,
        # The default timeouts should have been more than enough, but in some
        # extreme cases with a very slow debug build running on a slow or very busy
        # machine, they may not be. Observed tests reach 160 seconds. So it's
        # incremented to 200 seconds.
        # See issue #11289.
        # NOTE: request_timeout is the main cause of timeouts, even if logs say heartbeat
        request_timeout=200)
    whitelist_profile = ExecutionProfile(
        load_balancing_policy=TokenAwarePolicy(WhiteListRoundRobinPolicy(hosts)),
        consistency_level=ConsistencyLevel.LOCAL_QUORUM,
        serial_consistency_level=ConsistencyLevel.LOCAL_SERIAL,
        request_timeout=200)
    if use_ssl:
        # Scylla does not support any earlier TLS protocol. If you try,
        # you will get mysterious EOF errors (see issue #6971) :-(
        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
    else:
        ssl_context = None

    return Cluster(execution_profiles={EXEC_PROFILE_DEFAULT: profile, 'whitelist': whitelist_profile},
                   contact_points=hosts,
                   port=port,
                   # TODO: make the protocol version an option, to allow testing with
                   # different versions. If we drop this setting completely, it will
                   # mean pick the latest version supported by the client and the server.
                   protocol_version=4,
                   # NOTE: No auth provider as auth keysppace has RF=1 and topology will take
                   # down nodes, causing errors. If auth is needed in the future for topology
                   # tests, they should bump up auth RF and run repair.
                   ssl_context=ssl_context,
                   # The default timeouts should have been more than enough, but in some
                   # extreme cases with a very slow debug build running on a slow or very busy
                   # machine, they may not be. Observed tests reach 160 seconds. So it's
                   # incremented to 200 seconds.
                   # See issue #11289.
                   connect_timeout = 200,
                   control_connection_timeout = 200,
                   # NOTE: max_schema_agreement_wait must be 2x or 3x smaller than request_timeout
                   # else the driver can't handle a server being down
                   max_schema_agreement_wait=20,
                   idle_heartbeat_timeout=200,
                   # The default reconnection policy has a large maximum interval
                   # between retries (600 seconds). In tests that restart/replace nodes,
                   # where a node can be unavailable for an extended period of time,
                   # this can cause the reconnection retry interval to get very large,
                   # longer than a test timeout.
                   reconnection_policy = ExponentialReconnectionPolicy(1.0, 4.0),

                   auth_provider=auth_provider,
                   # Capture messages for debugging purposes.
                   connection_class=CustomConnection
                   )
```

BTW why do we need a **yet another** method to connect to clusters?

**Comment by kbr-scylla** [comment]
Opened https://github.com/scylladb/scylladb/issues/17914

**Comment by xtrey** [comment]
@kbr-scylla you are right. There is no need for an additional method. The fix is there https://github.com/scylladb/scylladb/pull/17932

**Comment by nyh** [comment]
I just fell on the cassandra-rackdc.properties formatting bug when backporting some test to an older branch (https://github.com/scylladb/scylla-enterprise/pull/4934).

**Pull request closed without merging**
